server/pulp/plugins/file/model_distributor.py

```diff
@@ -0,0 +1,249 @@
+from gettext import gettext as _
+import csv
+import errno
+import logging
+import os
+import shutil
+import traceback
+
+from pulp.common.plugins.distributor_constants import MANIFEST_FILENAME
+from pulp.plugins.distributor import Distributor
+from pulp.plugins.file.distributor import FilePublishProgressReport
+from pulp.server.controllers import repository as repo_controller
+from pulp.server.db import model
+from pulp.server.managers.repo import _common as common_utils
+
+
+BUILD_DIRNAME = 'build'
+
+_logger = logging.getLogger(__name__)
+
+
+class FileDistributor(Distributor):
+    """
+    Distribute files on the filesystem.
+    """
+
+    def __init__(self):
+        super(FileDistributor, self).__init__()
+        # Initialize instance variables used for writing out the metadata
+        self.metadata_file = None
+        self.metadata_csv_writer = None
+
+    @classmethod
+    def metadata(cls):
+        """
+        Advertise the capabilities of the mighty FileDistributor.
+
+        :return: The description of the impressive FileDistributor's capabilities.
+        :rtype:  dict
+        """
+        raise NotImplementedError()
+
+    def distributor_removed(self, repo, config):
+        """
+        Delete the published files from our filesystem
+
+        Please also see the superclass method definition for more documentation on this method.
+
+        :param repo: metadata describing the repository
+        :type  repo: pulp.server.db.model.Repository
+
+        :param config: plugin configuration
+        :type  config: pulp.plugins.config.PluginCallConfiguration
+        """
+        self.unpublish_repo(repo, config)
+
+    def publish_repo(self, repo, publish_conduit, config):
+        """
+        Publish the repository.
+
+        :param repo:            metadata describing the repo
+        :type  repo:            pulp.server.db.model.Repository
```

Comment:
Actually how can this work? The caller would need to also switch its type, but I don't see where that is done in this PR. Also when I run this code with the mongoengine-converted puppet branch I see repo is typed as `pulp.server.db.model.Repository`.

In my stack trace the caller into publish_repo comes from [_do_publish()](https://github.com/bmbouter/pulp/blob/18430255f42e854549fe684a2d19e8ea149e6feb/server/pulp/server/controllers/repository.py#L766) which does have both the `repo_obj` and `transfer_repo`. I think switching it will cause the type to switch for other polymorphic distributors so I'm not sure we can actually change this now. What do you think?